### ISSUE TITLE
Flags support

### ISSUE THREAD

**Issue by googleson78** [Parasrah/elm-react-component#40]
Is there support for passing elm flags, that I'm not finding?

**Comment by Parasrah** [comment]
No there is no support for flags, although it could potentially be done so the initial props are sent in as flags. Just to help understand, what is the use case for this?

**Comment by googleson78** [comment]
Hi, thanks for the quick response!

My use case is that I have an elm app that I want to control the url - it's almost the entire website, the only react parts are a header and footer.

Currently I use `Browser.application`, whose `init` gets the current url as an argument. `Browser.element` however, does not. Therefore my immediate first idea is to pass the url in from react as a flag.

**Comment by Parasrah** [comment]
Working within the constraints of how this library currently works, you should be able to pass the url as a React prop to the Elm component.

I do think it would be nice to update this library to support receiving the initial props as flags too to avoid having to define initial state in both React and Elm. I'll try to find some time at some point to do some profiling to evaluate if flags vs ports for initial render makes a visible difference for users

**Comment by googleson78** [comment]
I'm definitely missing something here - how can I have access to the url in my `init` when passing it in as a prop/port? Or do you mean to define a `port` and immediately pass data into it on startup, without doing anything in `init`?

**Comment by Parasrah** [comment]
Sorry I think I might be misunderstanding, is your intention to use `Browser.element` or `Browser.application`? If you just need to be able to store the url in your model using `Browser.element`, what I was suggesting was pass it in as a prop from the react side, and define an elm port to receive the url for your elm element. I could definitely be missing something though

**Comment by googleson78** [comment]
I need to use `Browser.element`, since I want to embed my elm app in a react project.

I am already passing the url in from a prop, as a port, and storing it in my `Model`. However I have the issue that at the point at which `Browser.element.init` runs there is no `Model` yet! (indeed, it creates the initial model) Therefore I have no access to the url prop at that point - you only have access to flags, which is why I want to try passing it in as a flag.

**Comment by Parasrah** [comment]
Ok that makes sense. In that case I would actually suggest taking a look at https://github.com/cultureamp/react-elm-components as I honestly think it would be a better fit for your use-case. My intention for this library was to make it so people could experiment with Elm with something really simple so they could get a feel for it, and as such it makes some pretty opinionated decisions to make things simpler. If it was me and I was building an app from scratch, and wanted the majority of it to be written in Elm, I'd use Evan's library because it's more flexible for more complicated scenarios like this. I think there are situations where this library is a better fit, just this isn't one of them.

**Comment by googleson78** [comment]
I'm actively avoiding it as it seems to be unmaintained (+old), and [not supporting newer react versions](https://github.com/cultureamp/react-elm-components/issues/18). (i.e. I managed to get this working, and not that library)

Thanks for the info/responses!

(btw, just my 5 cents - aren't flags even simpler than ports? i.e. if opinionated ports fit in "simple it just works" style, I think flags would be an even better fit - the dead-simple way I can think of passing them to elm is like in vanilla js, as some kind of `flags` prop)

**Comment by Parasrah** [comment]
I didn't realize that it was having issues with maintenance. You're totally right, it should be easier to do ports than flags. It just starts introducing a lot of edge cases depending on the implementation

For example if you could pass in a `flags` prop, that would conflict with anyone that wants to actually pass a prop called flags. That's probably an acceptable trade-off because it would be confusing to have a port called flags anyways, but it is something to consider.

Personally I think if one could just pass in the props as they do now, and access the initial state of those props as flags, it would be the simplest for people to use. That comes with its own potential edge cases (ports with no flags, flags with no ports, etc) although that would probably be trivial to work around.

I would like to work on this at some point, although I don't think I'm going to have time in the next few weeks. If the easiest way forward is using this library for you, feel free to fork to make it work for you, it's pretty small so it should be easy enough. If you keep me in the loop and it's the right solution for the lib we can merge it back upstream.

**Comment by dtaskoff** [comment]
I might be missing something, but it seems to me that passing the flags as something different than a `flags` prop isn't possible with the current implementation, because there is no way to differentiate between a flag, and a port of the type `: (a -> msg) -> Sub msg`.
Take as an example the following:
```js
<Main
  prop1={prop1}
  prop2={prop2}
  prop3={(arg) => { f(arg) }
/>
```

Clearly, prop3 is a port, but what about prop1, and prop2? Are they flags, or ports? For which of them should we call `sendData` on updates?
I would stick with passing a `flags` prop, although, as @Parasrah said, it comes with the cost, that it won't be possible to pass a port called `flags` (which I don't think is bad anyway).

**Comment by Parasrah** [comment]
I felt it should be both flags and ports, to be more in line with how React works. React doesn't have separate concepts for how to initialize a component and how to update it, whereas Elm does. If we treat props as both flags and ports, the Elm side is free to ignore either the flags or the ports if they don't need them.

The main purpose of this library is to make it easier to incrementally introduce Elm into an existing codebase, so that individuals and companies can try it out. That's why I'd prefer to conform to the react way of doing things if possible.

That said I'm definitely open to discussion about it, if you guys think it will make it harder to understand Elm for a beginner by blending the two concepts then maybe a `flags` prop would make more sense

**Comment by Parasrah** [comment]
I may try to make some time to fix up the tests on your PR and get it into master so that people can make use of this feature (since it's a common use case)

**Comment by googleson78** [comment]
If this is the route that you want to take, we should also add an check if the corresponding port exists before `sendData`ing it, because right now you can't define a flag without a corresponding port (because it blows up, because the corresponding elm port doesn't exist when you try to `sendData` to it)

**Comment by Parasrah** [comment]
That's a really good point I hadn't considered. If you two don't mind me asking @googleson78 @dtaskoff what is your use case for this library? Are you new to either Elm or React? And will there be team members using it whose experience differs from your own? If you guys think I'm off track with treating props as ports & flags to be easier for React users, I'm definitely open to your suggestions

**Comment by dtaskoff** [comment]
We aren't new to neither Elm, nor React, however, there are some team members being more familiar with Elm, and others being more familiar with React. Our team wrote a web service in Elm, but it turned out that we must reuse some large React components. So, we built a React application, and we plugged our web service into it.
As for the props, I don't think that treating them either way is easier for React, or Elm users.
It's just that flags are used only when initialisating an Elm element/application, while ports are used for continuous communication between Elm, and JavaScript. As an example, passing ports, and flags, like they are the same thing (like in `<Elm myFlag={fvalue} myPort={pvalue} />`), might mislead people into thinking, that sending data through a port will update the component.
I don't know, I don't have a strong preference, especially when there's good documentation.
If you have any other questions, I'll be happy to answer them.

**Comment by googleson78** [comment]
I have close to zero experience with React, and little with Elm.

I think I'm in favour of just reserving the `flags` name, as it is both simple and straightforward. The other approach of always sending the props as flags has some subtleties (e.g. don't `sendData` on non-existing ports, but still warn the user that he possibly forgot a port). I also think the loss of the `flags` name for a port isn't too bad, seeing as how the word already has the same meaning in Elm land, i.e. it would be a bad idea to name your port `flags` anyway, because of possible confusion between the two.

Have you had experience with people who usually write React, but want to try Elm, confusing the two concepts? I don't think someone that is willingly introducing Elm will have too many issues with flags as a concept, because it's the first thing in the Elm tutorial regarding js interop - and you need to read that part to understand ports anyway.

**Comment by Parasrah** [comment]
Those are all good points, and I agree I think if someone is naming a prop `flags` and passing it into an Elm element, it probably isn't the best practice in the first place. I also agree that Elm has a really clean division between initialization and updates.

> It's just that flags are used only when initializing an Elm element/application, while ports are used for continuous communication between Elm, and JavaScript

I think this is a good way to describe the situation from the perspective of an Elm developer. Meanwhile, I think the perspective from a React developer is something more like

"Props are something that are used to feed data into a component"

When you pass a prop into a react component, it will both initialize the component with that data, and cause that component to update when the data changes. Forgetting about the fact that the React component has anything to do with Elm, I think it would be confusing to most React developers to have one prop to initialize the component, and a different prop to update it.

I initially built this library in the hopes that one day I can introduce Elm where I work. So I want the API as friendly as possible for people that have a solid understanding of the React side as most of them have never been exposed to Elm before. That said I think I should sit down and figure out exactly how that would work while preserving the nice error messages and everything else

>  As an example, passing ports, and flags, like they are the same thing (like in `<Elm myFlag={fvalue} myPort={pvalue} />`), might mislead people into thinking, that sending data through a port will update the component

I think I might be misunderstanding this example. Do you mean if the ports on the Elm side haven't been hooked up, people might assume the Elm element will get updated data when it won't? If so that's a good point and I could definitely see that happening. Like you said maybe this can be addressed in a `FAQ` section in the docs or potentially another solution.

**Comment by Parasrah** [comment]
Sorry this took an incredibly long time to prioritize, but it has now been implemented in version `2.0.0`. I originally did attempt to do things the way that was suggested here, but quickly realized the implications of that. Having a port and a flag named the same just meant the flag would immediately be overwritten by the port. I was confused by this at first, so I can only imagine how confusing it would be for someone who didn't write the library. As such I decided on passing the props in as flags as well, you can see usage on the README
